protocols/imap/src/main/java/org/apache/james/imap/processor/base/UidMsnConverter.java

```diff
@@ -0,0 +1,107 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.imap.processor.base;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+
+import org.apache.james.mailbox.MessageUid;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Optional;
+import com.google.common.collect.Lists;
+
+public class UidMsnConverter {
+
+    public static final int FIRST_MSN = 1;
+
+    @VisibleForTesting final ArrayList<MessageUid> uids;
+
+    public UidMsnConverter(Iterator<MessageUid> iterator) {
+        uids = Lists.newArrayList(iterator);
+        Collections.sort(uids);
+    }
+
+    public synchronized Optional<Integer> getMsn(MessageUid uid) {
+        Collections.sort(uids);
```

Comment:
Ok, I think there are some cases which this could lead to a bug; but they are really extreme cases:
* max uid reached -> go back to 1
* two new concurrent uids, with some remove arround...
as this part of code is just giving us information, the 0.000001% bugs are really acceptable.